test/unit/messaging/process-statement-message.test.js

```diff
@@ -1,23 +1,59 @@
 jest.mock('ffc-messaging')
+jest.mock('../../../app/data')
+const mockGenerator = jest.fn()
+jest.mock('../../../app/generator', () => {
+  return mockGenerator
+})
+const mockValidation = jest.fn()
+jest.mock('../../../app/messaging/validate-statement', () => {
+  return mockValidation
+})
 const processStatementMessage = require('../../../app/messaging/process-statement-message')
+const mockStatement = require('../../mock-statement-data')
+const { VALIDATION } = require('../../../app/errors')
 let receiver
 
 describe('process statement message', () => {
   beforeEach(() => {
     receiver = {
-      completeMessage: jest.fn()
+      completeMessage: jest.fn(),
+      deadLetterMessage: jest.fn()
     }
   })
 
   afterEach(() => {
     jest.clearAllMocks()
   })
 
-  test('completes message', async () => {
+  test('completes message on success', async () => {
     const message = {
-      body: {}
+      body: mockStatement
     }
     await processStatementMessage(message, receiver)
     expect(receiver.completeMessage).toHaveBeenCalledWith(message)
   })
+
+  test('does not complete message on error', async () => {
+    const message = {
+      body: mockStatement
+    }
+    mockGenerator.mockImplementation(() => { throw new Error('Unable to generate') })
+    await processStatementMessage(message, receiver)
+    expect(receiver.completeMessage).not.toHaveBeenCalledWith(message)
+    expect(receiver.deadLetterMessage).not.toHaveBeenCalledWith(message)
+  })
+
+  test('dead letters message if validation error', async () => {
+    const message = {
+      body: mockStatement
+    }
+    mockValidation.mockImplementation(() => {
+      const err = new Error('Validation error')
+      err.category = VALIDATION
+      throw err
+    })
+    await processStatementMessage(message, receiver)
+    expect(receiver.completeMessage).not.toHaveBeenCalledWith(message)
+    expect(receiver.deadLetterMessage).toHaveBeenCalledWith(message)
+  })
```

Comment:
need test cases for:

- does it call `validateStatement`?
- how many times does it call `validateStatement`?
- what does it call `validateStatement` with?
- what if `validateStatement` throws?
- does it call `generateStatement`?
- how many times does it call `generateStatement`?
- what does it call `generateStatement` with?
- what if `generateStatement` throws?
- how many times does it call `receiver.completeMessage`?
- what if `receiver.completeMessage` throws?
- how many times does it call `receiver.deadLetterMessage`?
- what if `receiver.deadLetterMessage` throws?

Reply:
Added tests to cover the majority of these.

`deadLetterMessage` and `completeMessage` are a bit tricky as they're Service Bus functions.  So the actual behaviour of those failing is handled by the Azure SDKs.  They will send a message back to Service Bus to state to log the processing failure.  There's no real way for us to test that, particularly as we're mocking the SDK in the unit tests.